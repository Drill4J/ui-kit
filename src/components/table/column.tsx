import { ColumnProps } from './table-types';

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export const Column = (props: ColumnProps) => null;
