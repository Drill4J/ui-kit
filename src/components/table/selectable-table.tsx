import * as React from 'react';

import { Inputs } from '../../forms';
import { Table } from './table';
import { Column } from './column';

interface Props {
  data: any[];
  children: React.ReactNode;
  idKey: string;
  footer?: React.ReactNode;
  selectedRows: string[];
  onSelect: (selectedItems: string[]) => any;
  columnsSize?: 'wide' | 'medium';
  checkboxDescriptor?: <O>(item: O) => boolean;
  defaultSortField?: string;
}

export const SelectableTable = ({
  children,
  data,
  onSelect,
  idKey,
  selectedRows,
  checkboxDescriptor,
  defaultSortField,
  ...restProps
}: Props) => {
  const isAllSelected = selectedRows.length === data.length;

  return (
    <Table data={data} selectedRows={selectedRows} idKey={idKey} {...restProps} defaultSortField={defaultSortField}>
      <Column
        name="selector"
        Cell={({ item }) => {
          const isSelectable = checkboxDescriptor ? checkboxDescriptor(item) : true;
          return isSelectable ? (
            <Inputs.Checkbox
              onChange={() => {
                selectedRows.includes(item[idKey])
                  ? onSelect(selectedRows.filter((selectedItem) => selectedItem !== item[idKey]))
                  : onSelect([...selectedRows, item[idKey]]);
              }}
              checked={isAllSelected || selectedRows.includes(item[idKey])}
            />
          ) : null;
        }}
        HeaderCell={() => (
          <Inputs.Checkbox
            onChange={() => {
              onSelect(
                !isAllSelected
                  ? data.reduce((acc, item: any) => {
                    if (!checkboxDescriptor) {
                      return [...acc, String(item[idKey])];
                    }
                    return checkboxDescriptor(item) ? [...acc, String(item[idKey])] : acc;
                  }, [])
                  : [],
              );
            }}
            checked={isAllSelected}
          />
        )}
      />
      {children}
    </Table>
  );
};
