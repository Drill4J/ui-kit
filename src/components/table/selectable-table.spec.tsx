import React from 'react';
import renderer from 'react-test-renderer';

import { SelectableTable } from './selectable-table';
import { Column } from './column';

describe('SelectableTable', () => {
  it('should match snapshot', () => {
    const tree = renderer
      .create(
        <SelectableTable
          data={[{
            funcDefinition: 'function sum(a, b) { return a + b; };',
            funcExpression: 'const sum = function sum(a, b) { return a + b; };',
            anonymousFuncExpression: 'const sum = function(a, b) { return a + b; };',
            lambdaFunction: 'const sum = (a, b) => { return a + b; };',
            arrowFunction: 'const sum = (a, b) => (a + b);',
          }]}
          idKey="selectableTable"
          selectedRows={
            ['lambdaFunction']
          }
          onSelect={() => {}}
        >
          <Column
            name="alias"
            label="Name"
            Cell={({ value, item: { buildVersion } }) => (
              <div
                onClick={() => {
                }}
              >
                {value || buildVersion}
              </div>
            )}
          />
          <Column
            name="addedDate"
            label="Added"
            Cell={({ value }) => <span>{(value)}</span>}
          />
          <Column
            name="totalMethods"
            HeaderCell={() => (
              <div>Total Methods</div>
            )}
          />
          <Column
            name="newMethods"
            HeaderCell={() => (
              <div>New</div>
            )}
          />
        </SelectableTable>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
