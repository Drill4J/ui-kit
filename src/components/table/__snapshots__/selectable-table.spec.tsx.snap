// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectableTable should match snapshot 1`] = `
<table
  className="table"
>
  <thead
    className="table__header"
  >
    <tr>
      <th
        style={
          Object {
            "width": undefined,
          }
        }
      >
        <div
          className="checkbox checkbox--checked"
          onClick={[Function]}
        >
          <input
            checked={true}
            className="checkbox__checkbox-input checkbox__checkbox-input--checked"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            type="checkbox"
            value=""
          />
          <div
            className="checkbox__checkbox-icon-wrapper"
            data-modifiers=""
          >
            <svg
              className="generic-icon checkbox__check-mark-icon"
              data-modifiers=""
              fillRule="evenodd"
              height={7}
              style={
                Object {
                  "WebkitTransform": "rotate(0deg)",
                }
              }
              transform="rotate(0)"
              viewBox="0 0 13 9"
              width={10}
            >
              <path
                d="M.727 3.821a.935.935 0 0 0 .001 1.376l4.294 4.106a.99.99 0 0 0 1.417-.04l6.872-7.358c.38-.408.36-1.045-.038-1.417l.204.19a.982.982 0 0 0-1.398.057L6.42 6.932a.968.968 0 0 1-1.394.038L1.973 4.03a1.068 1.068 0 0 0-1.45-.018l.204-.19z"
              />
            </svg>
          </div>
        </div>
      </th>
      <th
        style={
          Object {
            "width": undefined,
          }
        }
      >
        Name
      </th>
      <th
        style={
          Object {
            "width": undefined,
          }
        }
      >
        Added
      </th>
      <th
        style={
          Object {
            "width": undefined,
          }
        }
      >
        <div>
          Total Methods
        </div>
      </th>
      <th
        style={
          Object {
            "width": undefined,
          }
        }
      >
        <div>
          New
        </div>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      className="table__row"
    >
      <td>
        <div
          className="checkbox checkbox--checked"
          onClick={[Function]}
        >
          <input
            checked={true}
            className="checkbox__checkbox-input checkbox__checkbox-input--checked"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            type="checkbox"
            value=""
          />
          <div
            className="checkbox__checkbox-icon-wrapper"
            data-modifiers=""
          >
            <svg
              className="generic-icon checkbox__check-mark-icon"
              data-modifiers=""
              fillRule="evenodd"
              height={7}
              style={
                Object {
                  "WebkitTransform": "rotate(0deg)",
                }
              }
              transform="rotate(0)"
              viewBox="0 0 13 9"
              width={10}
            >
              <path
                d="M.727 3.821a.935.935 0 0 0 .001 1.376l4.294 4.106a.99.99 0 0 0 1.417-.04l6.872-7.358c.38-.408.36-1.045-.038-1.417l.204.19a.982.982 0 0 0-1.398.057L6.42 6.932a.968.968 0 0 1-1.394.038L1.973 4.03a1.068 1.068 0 0 0-1.45-.018l.204-.19z"
              />
            </svg>
          </div>
        </div>
      </td>
      <td>
        <div
          onClick={[Function]}
        />
      </td>
      <td>
        <span />
      </td>
      <td />
      <td />
    </tr>
  </tbody>
</table>
`;
